src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs

```diff
@@ -538,57 +539,67 @@ public void Reset()
                 cancellationToken);
         }
 
-        // These methods are for chunked http responses that use GetMemory/Advance
         private Memory<byte> GetChunkedMemory(int sizeHint)
         {
-            // The max size of a chunk will be the size of memory returned from the PipeWriter (today 4096)
-            // minus 5 for the max chunked prefix size and minus 2 for the chunked ending, leaving a total of
-            // 4089.
+            var originalSizeHint = sizeHint;
+
+            // To guarantee that sizeHint is a minimumSize, we need to call GetMemory with a sizeHint
+            // larger than what was provided. We add the begin and end chunk lengths here.
+            if (sizeHint > 0)
+            {
+                sizeHint += ChunkWriter.GetBeginChunkByteCount(sizeHint) + EndChunkLength;
+            }
 
             if (!_currentChunkMemoryUpdated)
             {
-                // First time calling GetMemory
-                _currentChunkMemory = _pipeWriter.GetMemory(sizeHint);
-                _currentChunkMemoryUpdated = true;
+                UpdateCurrentChunkMemory(sizeHint);
             }
 
-            var memoryMaxLength = _currentChunkMemory.Length - BeginChunkLengthMax - EndChunkLength;
-            if (_advancedBytesForChunk >= memoryMaxLength - sizeHint && _advancedBytesForChunk > 0)
+            if (_advancedBytesForChunk >= _currentChunkMemory.Length - _currentMemoryPrefixBytes - EndChunkLength - originalSizeHint && _advancedBytesForChunk > 0)
             {
-                // Chunk is completely written, commit it to the pipe so GetMemory will return a new chunk of memory.
                 var writer = new BufferWriter<PipeWriter>(_pipeWriter);
                 WriteCurrentMemoryToPipeWriter(ref writer);
                 writer.Commit();
-
                 _unflushedBytes += writer.BytesCommitted;
-                _currentChunkMemory = _pipeWriter.GetMemory(sizeHint);
-                _currentChunkMemoryUpdated = true;
+
+                UpdateCurrentChunkMemory(sizeHint);
```

Comment:
sizeHint could still be 0 here right? What if the underlying pipe only gives you 5 bytes? I think that actualMemory could end up a length 0 bytes. If the underlying pipe gave less than 5 bytes, I think the slice would fail.